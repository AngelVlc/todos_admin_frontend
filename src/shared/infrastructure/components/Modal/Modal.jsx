import React, { useRef, forwardRef, useImperativeHandle } from "react";

export const Modal = forwardRef((props, ref) => {
  const modalRef = useRef();

  useImperativeHandle(ref, () => ({
    showModal() {
      toggleIsActive();
    },
    closeModal() {
      toggleIsActive();
    },
  }));

  const onCancel = () => {
    toggleIsActive();
  };

  const onOk = () => {
    props.closeHandler();
  };

  const toggleIsActive = () => {
    modalRef.current.classList.toggle("is-active");
  };

  return (
    <div ref={modalRef} className="modal" data-testid="modal">
      <div className="modal-background"></div>
      <div className="modal-content">
        <div className="box">
          <div className="content">{props.children}</div>
          <div className="field is-grouped is-justify-content-center">
            {props.showOk && (
              <div className="control">
                <button
                  className="button"
                  onClick={() => onOk()}
                  data-testid="modalOk"
                >
                  OK
                </button>
              </div>
            )}
            <div className="control">
              <button
                className="button"
                onClick={() => onCancel()}
                data-testid="modalCancel"
              >
                CANCEL
              </button>
            </div>
          </div>
        </div>
      </div>
      <button
        className="modal-close is-large"
        aria-label="close"
        data-testid="modalClose"
        onClick={() => onCancel()}
      ></button>
    </div>
  );
});
