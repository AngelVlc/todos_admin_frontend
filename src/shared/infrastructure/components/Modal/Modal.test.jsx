import React from "react";
import { render, cleanup, fireEvent, waitFor } from "@testing-library/react";
import { Modal } from "./Modal";
import { act } from "react-dom/test-utils";

const modalRef = React.createRef();

const mockedOnModalClose = jest.fn();

const renderModal = () => {
  return {
    ...render(
      <Modal ref={modalRef} closeHandler={mockedOnModalClose} showOk={true}>
        <p>Modal content</p>
      </Modal>
    ),
  };
};

afterEach(cleanup);

describe("Modal", () => {
  describe("When the modal is hidden", () => {
    it("should match the snapshot", async () => {
      let fragment;
      await act(async () => {
        const { asFragment } = renderModal();
        fragment = asFragment;
      });

      expect(fragment()).toMatchSnapshot();
    });
  });

  describe("When the modal is visible", () => {
    it("should match the snapshot", async () => {
      let fragment;
      await act(async () => {
        const { asFragment } = renderModal();
        fragment = asFragment;

        modalRef.current.showModal();
      });

      expect(fragment()).toMatchSnapshot();
    });
  });

  describe("Close button", () => {
    it("should close the modal without triggering the close handler", async () => {
      let container;
      await act(async () => {
        container = renderModal();

        modalRef.current.showModal();
      });

      await waitFor(() => {
        fireEvent.click(container.getByTestId("modalClose"));
      });

      expect(mockedOnModalClose).not.toHaveBeenCalled();
    });
  });

  describe("Close button", () => {
    it("should close the modal without triggering the close handler", async () => {
      let container;
      await act(async () => {
        container = renderModal();

        modalRef.current.showModal();
      });

      await waitFor(() => {
        fireEvent.click(container.getByTestId("modalClose"));
      });

      expect(mockedOnModalClose).not.toHaveBeenCalled();
    });
  });

  describe("Cancel button", () => {
    it("should close the modal without triggering the close handler", async () => {
      let container;
      await act(async () => {
        container = renderModal();

        modalRef.current.showModal();
      });

      await waitFor(() => {
        fireEvent.click(container.getByTestId("modalCancel"));
      });

      expect(mockedOnModalClose).not.toHaveBeenCalled();
    });
  });

  describe("Ok button", () => {
    it("should close the modal triggering the close handler", async () => {
      let container;
      await act(async () => {
        container = renderModal();

        modalRef.current.showModal();
      });

      await waitFor(() => {
        fireEvent.click(container.getByTestId("modalOk"));
      });

      expect(mockedOnModalClose).toHaveBeenCalled();
    });
  });

  describe("ref #closeModal", () => {
    it("should close the modal", async () => {
      let container;
      await act(async () => {
        container = renderModal();

        modalRef.current.showModal();
        modalRef.current.closeModal();
      });

      expect(container.getByTestId("modal").classList.contains('is-active')).toBe(false)
    })
  });
});
