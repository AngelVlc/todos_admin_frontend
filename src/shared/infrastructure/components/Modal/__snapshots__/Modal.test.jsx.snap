// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal When the modal is hidden should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="modal"
    data-testid="modal"
  >
    <div
      class="modal-background"
    />
    <div
      class="modal-content"
    >
      <div
        class="box"
      >
        <div
          class="content"
        >
          <p>
            Modal content
          </p>
        </div>
        <div
          class="field is-grouped is-justify-content-center"
        >
          <div
            class="control"
          >
            <button
              class="button"
              data-testid="modalOk"
            >
              OK
            </button>
          </div>
          <div
            class="control"
          >
            <button
              class="button"
              data-testid="modalCancel"
            >
              CANCEL
            </button>
          </div>
        </div>
      </div>
    </div>
    <button
      aria-label="close"
      class="modal-close is-large"
      data-testid="modalClose"
    />
  </div>
</DocumentFragment>
`;

exports[`Modal When the modal is visible should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="modal is-active"
    data-testid="modal"
  >
    <div
      class="modal-background"
    />
    <div
      class="modal-content"
    >
      <div
        class="box"
      >
        <div
          class="content"
        >
          <p>
            Modal content
          </p>
        </div>
        <div
          class="field is-grouped is-justify-content-center"
        >
          <div
            class="control"
          >
            <button
              class="button"
              data-testid="modalOk"
            >
              OK
            </button>
          </div>
          <div
            class="control"
          >
            <button
              class="button"
              data-testid="modalCancel"
            >
              CANCEL
            </button>
          </div>
        </div>
      </div>
    </div>
    <button
      aria-label="close"
      class="modal-close is-large"
      data-testid="modalClose"
    />
  </div>
</DocumentFragment>
`;
